components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/request/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,411 @@
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.authentication.framework.exception.PostAuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.handler.request.AbstractPostAuthnHandler;
+import org.wso2.carbon.identity.application.authentication.framework.handler.request.PostAuthnHandlerFlowStatus;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.IdentityProviderProperty;
+import org.wso2.carbon.identity.application.common.model.User;
+import org.wso2.carbon.identity.base.IdentityException;
+import org.wso2.carbon.identity.mgt.util.Utils;
+import org.wso2.carbon.identity.recovery.ChallengeQuestionManager;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryConstants;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryException;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryServerException;
+import org.wso2.carbon.identity.recovery.internal.IdentityRecoveryServiceDataHolder;
+import org.wso2.carbon.identity.recovery.model.ChallengeQuestion;
+import org.wso2.carbon.identity.recovery.model.UserChallengeAnswer;
+import org.wso2.carbon.idp.mgt.IdentityProviderManagementException;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.api.UserStoreManager;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class will check whether the challenge questions are set for the user.
+ * Also, It will force users to add answers to challenge questions is challenge questions are not
+ * already answered
+ **/
+
+public class PostAuthnMissingChallengeQuestionsHandler extends AbstractPostAuthnHandler {
+
+    private static final String CHALLENGE_QUESTIONS_REQUESTED = "challengeQuestionsRequested";
+    private static final String SELECTED_CHALLENGE_QUESTION_PREFIX = "Q-";
+    private static final String CHALLENGE_QUESTION_ANSWER_PREFIX = "A-";
+
+    private static final Log log = LogFactory.getLog(PostAuthnMissingChallengeQuestionsHandler.class);
+    private static volatile PostAuthnMissingChallengeQuestionsHandler instance;
+
+    public static PostAuthnMissingChallengeQuestionsHandler getInstance() {
+        if (instance == null) {
+            synchronized (PostAuthnMissingChallengeQuestionsHandler.class) {
+                if (instance == null) {
```

Comment:
redundant null check

Reply:
instead, double checked lock shall we initialize the variable during the declaration?

Reply:
yes, removed the double-checked lock and initialized during the declaration